### PR TITLE
[stdlib] [Arith] Cantor pairing

### PR DESCRIPTION
**Pull request by mrhaandi**
**Kind:** feature

This PR suggests to add the most rudimentary Cantor pairing function (and its inverse) to stdlib.
Along the lines of previous discussions (see [coq.discourse.group](https://coq.discourse.group/t/bijections-between-nat-and-nat-nat/720)), such a pairing function is the most fundamental building block of any countability argument (see [stdpp/contable.v](https://gitlab.mpi-sws.org/iris/stdpp/-/blob/master/theories/countable.v)).

The proposed PR is by design a very lightweight pairing implementation, providing basic functionality.
The provided `to_nat_non_decreasing` lemma is useful for size recursion wrt. the pairing (e.g. for implementing a bijection between `list nat` and `nat`).
Lemma `to_nat_spec` is the arithmetic [specification](https://en.wikipedia.org/wiki/Pairing_function#Cantor_pairing_function) of Cantor pairing. In the proof it would be possible to use only lemmas from `PeanoNat`, which is then somewhat cumbersome.
Therefore, `lia` and later `nia` tactics are used.

Along other possible pairing functions, this one is special because it is conjectured to be the only polynomial bijective pairing function.

For example, having this as part of stdlib would improve the infrastructrure of [coq-library-undecidability](https://github.com/uds-psl/coq-library-undecidability), as it relies on ad-hoc Cantor pairing for enumerability proofs.

Also, pairing allows for a generalization of `ConstructiveEpsilon` to semi-decidable predicates (currently, it is formulated for decidable ones).

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

File: doc/changelog/10-standard-library/14008-Cantor-pairing.rst
Comment (jfehrle) [inline doc/changelog/10-standard-library/14008-Cantor-pairing.rst:7]: Perhaps it would help the reader find your addition if you mention the file `Cantor.v`?  That it's in Arith seems less important.

My impression is that line 2 is sufficient as a summary of what's been added and that lines 3-7 might be better as comments in `Cantor.v` so that info can be easily found.  The changelog is just to bring things to people's attention.

WDYT?

File: theories/Arith/Cantor.v
Comment (mrhaandi) [inline theories/Arith/Cantor.v:62]: Do you know of any better solution to specify that `B` is even and  `to_nat (x, y) = B/2` such that `lia` and `nia` are not confused?

File: doc/changelog/10-standard-library/14008-Cantor-pairing.rst
Comment (mrhaandi) [reply doc/changelog/10-standard-library/14008-Cantor-pairing.rst:7]: Sure, I can make it less verbose.

File: theories/Arith/Cantor.v
Comment (olaure01) [reply theories/Arith/Cantor.v:70]: Not sure this is what you are looking for but `lia` deals well with `2 * _` apparently. So the following two specifications looks reasonable to have:
```coq
Lemma to_nat_spec x y :
  2 * to_nat (x, y) = 2 * y + (y + x) * (1 + y + x).
Proof.
  simpl. induction (y + x) as [|n IHn]; simpl; lia.
Qed.

Lemma to_nat_spec2 x y :
  to_nat (x, y) = y + (y + x) * (1 + y + x) / 2.
Proof.
  now rewrite <- Nat.div_add_l, Nat.mul_comm,
              <- to_nat_spec, <- Nat.div2_div, Nat.div2_double.
Qed.
```

Comment (mrhaandi) [reply theories/Arith/Cantor.v:70]: My main concern about `2 * _` is that upon `simpl` or `cbn` it is reduced to `_ + (_ + 0)` which feels strictly worse than `_ + _`.
Sometimes `_ * 2` worked better for me instead. I'm still undecided on either
- `to_nat (x, y) + to_nat (x, y)`
- `2 * to_nat (x, y)`
- `to_nat (x, y) * 2`
- `Nat.double (to_nat (x, y))`

Do you know of some pros/cons of the above possibilities?

Comment (mrhaandi) [reply theories/Arith/Cantor.v]: After some offline discussions I decided to use `to_nat (x, y) * 2` as it strikes a good balance between presentation, reduction behavior, and usability.
The proof of the newly added (as suggested) lemma `to_nat_spec2` also shows some small benefits compared to @olaure01's solution.

**Comment by mrhaandi** [comment]
Side note: with the proposed changes of https://github.com/coq/coq/pull/14037, for the proof of `to_nat_spec2` one can use `pose proof (to_nat_spec x y). lia.`.
I will keep the manual rewrites.

File: theories/Arith/Cantor.v
Comment (olaure01) [inline theories/Arith/Cantor.v:72]:
```suggestion
  to_nat (x, y) * 2 = y * 2 + (y + x) * (1 + y + x).
```

File: theories/Arith/Cantor.v
Comment (olaure01) [inline theories/Arith/Cantor.v:78]:
```suggestion
  to_nat (x, y) = y + (y + x) * (1 + y + x) / 2.
```

**Review by olaure01 — COMMENTED** [review]
Wouldn't it be interesting to integrate the coding of `list nat` in `nat` in this file?

**Comment by mrhaandi** [comment]
> Wouldn't it be interesting to integrate the coding of `list nat` in `nat` in this file?

I also considered this but decided not to.
This file is supposed to be (as part of Arith) about `nat`. Such a construction for `list` would fit more at the level alongside `List.v`.
Also, thought a little further, why stop at `list nat` and not have the same for e.g. `tree nat`, `Z`, ... or any infinite, enumerable type.
Maybe there will be demand for such methods in `stdlib` in future. For now, to avoid immediate feature creep and keep things clean, I prefer to have exactly the most basic building block given in this PR.

File: theories/Arith/Cantor.v
Comment (mrhaandi) [reply theories/Arith/Cantor.v:72]: The choice `(1 + (y + x))` is deliberate to have a common subterm `(y + x)` with the previous part (e.g. for potential rewrites) even before `cbn`.

File: theories/Arith/Cantor.v
Comment (mrhaandi) [reply theories/Arith/Cantor.v:78]: see reply to previous similar comment

File: theories/Arith/Cantor.v
Comment (olaure01) [reply theories/Arith/Cantor.v:72]:
```suggestion
  to_nat (x, y) * 2 = y * 2 + (y + x) * S (y + x).
```

File: theories/Arith/Cantor.v
Comment (olaure01) [reply theories/Arith/Cantor.v:78]:
```suggestion
  to_nat (x, y) = y + (y + x) * S (y + x) / 2.
```

**Review by olaure01 — APPROVED** [review]

**Comment by herbelin** [comment]
> Also, thought a little further, why stop at `list nat` and not have the same for e.g. `tree nat`, `Z`, ... or any infinite, enumerable type.

For the record,  [Qinxiang Cao and Xiwei Wu](https://github.com/QinxiangCao/Countable_ITP21) worked on a library  generating injections of simply-typed countable types into `nat`.

File: theories/Arith/Cantor.v
Comment (mrhaandi) [reply theories/Arith/Cantor.v:72]: Here I also thought back and forth and ultimately decided on `(1 + (y + x))`, for which `S (y + x)` is the normal form obtainable by `cbn` (if the user wishes so). `(1 + (y + x))` felt like the best mathematical presentation (with previously mentioned subterm) that can be reused verbatim for other natural number presentations without the specific constructor `S`.
The reason why `S (y + x)` could be better is _becase_ it is already the normal form for `nat`.
Do you have a strong reason why `S (y + x)` is definitely the better choice?

**Comment by mrhaandi** [comment]
> For the record, [Qinxiang Cao and Xiwei Wu](https://github.com/QinxiangCao/Countable_ITP21) worked on a library generating injections of simply-typed countable types into `nat`.

The `natnat_nat_bijection: bijection (prod nat nat)` lemma in the mentioned library could profit from this PR.

File: theories/Arith/Cantor.v
Comment (olaure01) [reply theories/Arith/Cantor.v:72]: No strong argument on my side, I am just still mentally blocking on this `1 + (y + x)`... possibly because I like statements which look as simple as possible. I perfectly understand your argument about being able to rewrite `y + x`. But I am still not sure there is some technical argument in favor of `1 + _` wrt `S _` (and using `S _` avoids the right-associated double sum). From the mathematical point of view, I would consider `1 + _` as an artifact w.r.t. the more primitive `S _` (but probably I would support the converse for a mathematical text...).

Comment (mrhaandi) [reply theories/Arith/Cantor.v:72]: I guess your suggestion is more in line with the style of stdlib (e.g. `PeanoNat.v`), so I'll use `S (y + x)`.

**Comment by olaure01** [comment]
@mrhaandi I think you can squash the commits, and I will merge on next Wednesday if there is no more comments.

**Comment by mrhaandi** [comment]
> @mrhaandi I think you can squash the commits, and I will merge on next Wednesday if there is no more comments.

Alright.

**Comment by olaure01** [comment]
@coqbot: merge now

**Pull request merged**
